arch/riscv/core/smp.c

```diff
@@ -74,11 +75,15 @@ void arch_secondary_cpu_init(int hartid)
 #endif
 #ifdef CONFIG_SMP
 	irq_enable(RISCV_IRQ_MSOFT);
+#endif
+#ifdef CONFIG_RISCV_SOC_HAS_CUSTOM_PER_CORE_INIT
+	z_soc_per_core_init();
 #endif
 	riscv_cpu_init[cpu_num].fn(riscv_cpu_init[cpu_num].arg);
 }
 
-#ifdef CONFIG_SMP
+/* IPI */
+#if defined(CONFIG_SMP) && !defined(CONFIG_RISCV_SOC_HAS_CUSTOM_SMP_IPI)
```

Comment:
nit [1/2]: I think the following makes it clearer that this portion of code is for `SMP`, and within it, there's a portion of code that shouldn't be compiled when `CONFIG_RISCV_SOC_HAS_CUSTOM_SMP_IPI=y`

```suggestion
#ifdef CONFIG_SMP
#ifndef CONFIG_RISCV_SOC_HAS_CUSTOM_SMP_IPI
```